### PR TITLE
add pulls to fetch owsk-openshift images locally to reduce wait times

### PR DESCRIPTION
**Pull request by evanchooly**
This change will prefetch the docker images required to deploy openwhisk reducing wait times during scenarios.

### CODE REVIEW

**Comment by evanchooly** [comment]
Excellent point.  Done.

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 1]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
 
 docker pull openshiftroadshow/parksmap-katacoda:1.0.0
+
+###  OpenWhisk images
+PROJECTODD_VERSION=f5eae82
+OPENWHISK_VERSION=rhdemo-b7724ef
+STRIMZI_VERSION=0.2.0
+docker pull busybox
+docker pull centos/nginx-112-centos7@sha256:42330f7f29ba1ad67819f4ff3ae2472f62de13a827a74736a5098728462212e7
+docker pull openwhisk/alarmprovider:1.9.0
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: change this to: `docker pull projectodd/alarmprovider:${PROJECTODD_LEARN_TAG}`

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 2]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
+docker pull projectodd/action-java-8:${PROJECTODD_VERSION}
+docker pull projectodd/action-nodejs-6:${PROJECTODD_VERSION}
+docker pull projectodd/action-nodejs-8:${PROJECTODD_VERSION}
+docker pull projectodd/action-php-7:${PROJECTODD_VERSION}
+docker pull projectodd/action-python-2:${PROJECTODD_VERSION}
+docker pull projectodd/action-python-3:${PROJECTODD_VERSION}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: 1,$s/PROJECTODD_VERSION/PROJECTODD_LEARN_TAG/g

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 3]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
+docker pull projectodd/controller:${OPENWHISK_VERSION}
+docker pull projectodd/invoker:${OPENWHISK_VERSION}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: 1,$s/OPENWHISK_VERSION/PROJECTODD_LEARN_TAG/g

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 4]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
+docker pull projectodd/whisk_alarms:${PROJECTODD_VERSION}
+docker pull projectodd/whisk_catalog:da00e0c
+docker pull projectodd/whisk_couchdb:${PROJECTODD_VERSION}
+docker pull strimzi/cluster-controller:${STRIMZI_VERSION}
+docker pull strimzi/kafka:${STRIMZI_VERSION}
+docker pull strimzi/zookeeper:${STRIMZI_VERSION}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: 1,$s/strimzi/projectodd/g

and

1,$s/STRIMZI_VERSION/PROJECTODD_LEARN_TAG/g

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 5]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
+docker pull projectodd/action-java-8:${PROJECTODD_VERSION}
+docker pull projectodd/action-nodejs-6:${PROJECTODD_VERSION}
+docker pull projectodd/action-nodejs-8:${PROJECTODD_VERSION}
+docker pull projectodd/action-php-7:${PROJECTODD_VERSION}
+docker pull projectodd/action-python-2:${PROJECTODD_VERSION}
+docker pull projectodd/action-python-3:${PROJECTODD_VERSION}
+docker pull projectodd/controller:${OPENWHISK_VERSION}
+docker pull projectodd/invoker:${OPENWHISK_VERSION}
+docker pull projectodd/whisk_alarms:${PROJECTODD_VERSION}
+docker pull projectodd/whisk_catalog:da00e0c
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: Change this to `docker pull projectodd/whisk_catalog:${PROJECTODD_LEARN_TAG}`

File: environments/openshift-middleware-3-7/scripts/pull.sh (hunk 2 above)
Comment (jimcurtis) [reply environments/openshift-middleware-3-7/scripts/pull.sh]: Also, need to add `+docker pull projectodd/action-swift-3:${PROJECTODD_LEARN_TAG}`

File: environments/openshift-middleware-3-7/scripts/pull.sh [hunk 6]
```diff
@@ -22,3 +22,27 @@ docker pull centos/python-35-centos7:latest
...
+docker pull projectodd/whisk_couchdb:${PROJECTODD_VERSION}
+docker pull strimzi/cluster-controller:${STRIMZI_VERSION}
+docker pull strimzi/kafka:${STRIMZI_VERSION}
+docker pull strimzi/zookeeper:${STRIMZI_VERSION}
+
+curl -o /opt/openwhisk-template https://git.io/openwhisk-template
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/pull.sh]: https://git.io/openwhisk-template currently points to projectodd/openwhisk-openshift/master/template.yml.  I am creating a new template that points to our stable katacoda/learn images.  It is called projectodd/openwhisk-openshift/master/learn-template.yml.  We should curl this new one instead.

**Comment by rafabene** [comment]
I think we should move the openwhisk commands to an openwhisk-pull.sh script just like istio-pull.sh

**Comment by evanchooly** [comment]
@rafabene already done :)

**Comment by rafabene** [comment]
Thanks.

I miss the following images on the script:

- strimzi/cluster-controller:latest              370 MB
- strimzi/kafka:latest                                403.7 MB
- strimzi/zookeeper:latest                       403.7 MB
- busybox:latest                                      1.146 MB
- openwhisk/alarmprovider:1.9.0            364.5 MB
- centos/nginx-112-centos7@sha256:42330f7f29ba1ad67819f4ff3ae2472f62de13a827a74736a5098728462212e7 325.1 MB

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+###  OpenWhisk images
+PROJECTODD_LEARN_TAG=learn
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: We need to change this back to `PROJECTODD_VERSION=f5eae82`

Then add back:

OPENWHISK_VERSION=rhdemo-b7724ef
STRIMZI_VERSION=0.2.0

Then, do:

:1,$s/PROJECTODD_LEARN_TAG/PROJECTODD_VERSION/g

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/busybox:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: change back to:

`docker pull busybox`

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/nginx:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: change back to:

docker pull centos/nginx-112-centos7@sha256:42330f7f29ba1ad67819f4ff3ae2472f62de13a827a74736a5098728462212e7

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/alarmprovider:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: change back to:

docker pull openwhisk/alarmprovider:1.9.0

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/action-java-8:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-nodejs-6:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-nodejs-8:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-php-7:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-python-2:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-python-3:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/action-swift-3:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/controller:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: change back to:

docker pull projectodd/controller:${OPENWHISK_VERSION}

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/invoker:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/whisk_alarms:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/whisk_catalog:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: change back to:

docker pull projectodd/whisk_catalog:da00e0c

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+docker pull projectodd/whisk_couchdb:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/cluster-controller:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/kafka:${PROJECTODD_LEARN_TAG}
+docker pull projectodd/zookeeper:${PROJECTODD_LEARN_TAG}
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: do:

:20,22s/projectodd/strimzi/g

and:

:20/22s/PROJECTODD_VERSION/STRIMZI_VERSION/g

File: environments/openshift-middleware-3-7/scripts/owsk-pull.sh [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+# leave both in for now until the learn template gets merged
+# the scenarios can reference /opt/openwhisk-template in the meantime
+curl -o /opt/openwhisk-template https://raw.githubusercontent.com/projectodd/openwhisk-openshift/master/template.yml
+curl -o /opt/openwhisk-learn-template https://raw.githubusercontent.com/projectodd/openwhisk-openshift/master/learn-template.yml
```
Comment (jimcurtis) [inline environments/openshift-middleware-3-7/scripts/owsk-pull.sh]: remove this curl of learn-template since we are not going to use it.

**Comment by evanchooly** [comment]
done and dusted.  ready to go.

File: environments/openshift-middleware-3-7/scripts/owsk-install.sh [hunk 15]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
```
Comment (rafabene) [inline environments/openshift-middleware-3-7/scripts/owsk-install.sh]: I don't see where this script is being used.

File: environments/openshift-middleware-3-7/scripts/owsk-install.sh [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+[ -f /usr/local/bin/wsk ] || sudo tar xzvf OpenWhisk_CLI-latest-linux-386.tgz -C /usr/local/bin wsk
+
+until $(oc status &> /dev/null); do sleep 1; done; oc adm policy add-cluster-role-to-user cluster-admin admin
+oc new-project faas --display-name="FaaS - Apache OpenWhisk"
+oc adm policy add-role-to-user admin developer -n faas
+
+oc process -f https://git.io/openwhisk-template | oc create -f -
+#oc process -f https://github.com/projectodd/openwhisk-openshift/master/learn-template.yml | oc create -f -
+
+git clone https://github.com/apache/incubator-openwhisk-devtools /tmp/openwhisk-devtools
```
Comment (rafabene) [inline environments/openshift-middleware-3-7/scripts/owsk-install.sh]: Shouldn't this project be also at https://github.com/openshift-labs/learn-katacoda/blob/master/environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh#L9 to pre-fetch the maven repo?

Comment (evanchooly) [reply environments/openshift-middleware-3-7/scripts/owsk-install.sh]: eventually that's the plan though there's some discussion around when/where that kind of image will get integrated.

File: environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh [hunk 17]
```diff
@@ -237,3 +237,9 @@ done
 popd > /dev/null
 echo - DONE | tee -a $SCRIPT_LOG
 
+git clone https://github.com/apache/incubator-openwhisk-devtools /tmp/openwhisk-devtools
+
+cd /tmp/openwhisk-devtools/java-action-archetype \
```
Comment (rafabene) [inline environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh]: To populate the maven repo, it should be on line *9: https://github.com/evanchooly/learn-katacoda/blob/master/environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh#L9 but it won't install the resources. 

The mvn clean

Comment (evanchooly) [reply environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh]: oh, ok.  i'll just completely revert that particular change then.

**Comment by rafabene** [comment]
@evanchooly Although you have included the https://github.com/apache/incubator-openwhisk-devtools in the owsk-install.sh script, this script is not executed on the environment creation. This means that any Maven dependency will only be resolved and fetched when you run the script for the first time. To avoid downloading the dependencies from Maven central, please add the repo https://github.com/apache/incubator-openwhisk-devtools to env var SOURCE_REPOS at https://github.com/openshift-labs/learn-katacoda/blob/master/environments/openshift-middleware-3-7/scripts/populate-maven-repos.sh#L9 - This will pre-fetch the Maven dependencies in the environment, making the openwhisk installation faster by avoiding downloading Maven dependencies.

**Comment by evanchooly** [comment]
@rafabene ah, i get it now.  just updated that script.

**Comment by rafabene** [comment]
I'm ok with this PR now if the scripts work well.

**Pull request merged**